File: 3/js/src/BoggleBoard.ts
import { TrieNode, Trie } from "./Trie";

function boggleBoard(board: string[][], words: string[]) {
	const r = board.length;
	const c = board[0].length;

	const visited: boolean[][] = board.map((row) => row.map(() => false));
	const res: Set<string> = new Set();

	const trie: Trie = new Trie();
	for (const word of words) {
		trie.insert(word);
	}

	for (let i = 0; i < r; i++) {
		for (let j = 0; j < c; j++) {
			visit(i, j, trie.root, board, visited, res);
		}
	}

	return res;
}

function visit(
	i: number,
	j: number,
	cur: TrieNode,
	board: string[][],
	visited: boolean[][],
	res: Set<string>
) {
	if (visited[i][j]) {
		return;
	}

	const char = board[i][j];

	if (!cur.children.has(char)) {
		return;
	}

	cur = cur.children.get(char)!;
	visited[i][j] = true;

	if (cur.children.has(Trie.END_SYMBOL)) {
		res.add(cur.word);
	}

	const neighbours = getUnvisitedNeighbours(i, j, board, visited);
	for (const neighbour of neighbours) {
		visit(neighbour[0], neighbour[1], cur, board, visited, res);
	}

	visited[i][j] = false;
}

function getUnvisitedNeighbours(
	i: number,
	j: number,
	board: string[][],
	visited: boolean[][]
): [number, number][] {
	const X = [0, 0, 1, -1, 1, -1, -1, 1];
	const Y = [1, -1, 0, 0, 1, -1, 1, -1];

	const unvisitedNeighbours: [number, number][] = [];
	for (let k = 0; k < 8; k++) {
		const x = X[k] + i;
		const y = Y[k] + j;

		if (
			x >= 0 &&
			x < board.length &&
			y >= 0 &&
			y < board[0].length &&
			!visited[x][y]
		) {
			unvisitedNeighbours.push([x, y]);
		}
	}

	return unvisitedNeighbours;
}

(function main() {
	const board = [
		["t", "h", "i", "s", "i", "s", "a"],
		["s", "i", "m", "p", "l", "e", "x"],
		["b", "x", "x", "x", "x", "e", "b"],
		["x", "o", "g", "g", "l", "x", "o"],
		["x", "x", "x", "D", "T", "r", "a"],
		["R", "E", "P", "E", "A", "d", "x"],
		["x", "x", "x", "x", "x", "x", "x"],
		["N", "O", "T", "R", "E", "-", "P"],
		["x", "x", "D", "E", "T", "A", "E"],
	];

	const words = [
		"this",
		"is",
		"not",
		"a",
		"simple",
		"boggle",
		"board",
		"test",
		"REPEATED",
		"NOTRE-PEATED",
	];

	console.log(boggleBoard(board, words));
})();

File: 3/java/BoggleBoard.java
import java.util.Set;
import java.util.HashSet;
import java.util.Map;
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;

class BoggleBoard {
	public static Set<String> boggleBoard(String words[], char board[][]){
		Trie trie = new Trie();
		for(String word: words) {
			trie.add(word);
		}
		
		int r = board.length;
		int c = board[0].length;
		boolean visited[][] = new boolean[r][c];
	
		Set<String> finalWords = new HashSet<>();
		
		for(int i=0; i<r; i++) {
			for(int j=0; j<c; j++) {
				explore(i, j, board, visited, trie.root, finalWords);
			}
		}

		return finalWords;
	}

	public static void explore(
			int i,
			int j,
			char board[][],
			boolean visited[][],
			TrieNode root,
			Set<String> finalWords
			) {
		if(visited[i][j]) {
			return;
		}
	
		char letter = board[i][j];

		if(!root.children.containsKey(letter)) {
			return;
		}
		
		visited[i][j] = true;
		root = root.children.get(letter);
		
		if(root.children.containsKey('*')) {
			finalWords.add(root.word);
		}	
		
		List<int[]> neighbours = getNeighbours(i, j, board);
		for(int[] neighbour : neighbours) {
			explore(neighbour[0], neighbour[1], board, visited, root, finalWords);
		}
		visited[i][j] = false;
	}
	
	public static List<int[]> getNeighbours(int i, int j, char board[][]) {
		List<int[]> neighbours =  new ArrayList<>();
		
		int xDir[] = {0, 1, 0, -1, 1, -1, 1, -1};
		int yDir[] = {1, 0, -1, 0, 1, -1, -1, 1};
			
		for(int k=0; k<xDir.length; k++) {
			int newI = i + xDir[k];
			int newJ = j + yDir[k]; 
			
			if( (newI >= 0 && newI < board.length) && (newJ >=0 && newJ < board[0].length)  ) {
				neighbours.add(new int[]{newI, newJ});
			}
		}

		return neighbours;
	}

	public static void main(String args[]) {
		String words[] = {"this", "is", "not", "a", "simple", "boggle", "board", "test", "REPEATED", "NOTRE-PEATED"};
		char board[][] = {
			{'t', 'h', 'i', 's', 'i', 's', 'a'},
			{'s', 'i', 'm', 'p', 'l', 'e', 'x'},
			{'b', 'x', 'x', 'x', 'x', 'e', 'b'},
			{'x', 'o', 'g', 'g', 'l', 'x', 'o'},
			{'x', 'x', 'x', 'D', 'T', 'r', 'a'},
			{'R', 'E', 'P', 'E', 'A', 'd', 'x'},
			{'x', 'x', 'x', 'x', 'x', 'x', 'x'},
			{'N', 'O', 'T', 'R', 'E', '-', 'P'},
			{'x', 'x', 'D', 'E', 'T', 'A', 'E'}
		};
		
		System.out.println(boggleBoard(words, board));
	}
}

class TrieNode {
	Map<Character, TrieNode> children;
	String word;

	TrieNode() {
		this.children = new HashMap<>();
		this.word = "";
	}
}

class Trie {
	TrieNode root;
	
	Trie() {
		this.root = new TrieNode();
	}
	
	void add(String word) {
		TrieNode node = root;

		for(char letter : word.toCharArray()) {
			if(!node.children.containsKey(letter)) {
				node.children.put(letter, new TrieNode());
			}
			
			node = node.children.get(letter);
		}	
		
		node.children.put('*', null);
		node.word = word;
	}
}
